packages/core/src/ui/UICanvas.ts

```diff
@@ -0,0 +1,561 @@
+    const { _renderCamera: renderCamera, _transform: transform } = this;
+    if (renderCamera) {
+      const { transform: cameraTransform } = renderCamera.entity;
+      const { worldPosition: cameraWorldPosition, worldForward: cameraWorldForward } = cameraTransform;
+      const { _distance: distance } = this;
+      transform.setWorldPosition(
+        cameraWorldPosition.x + cameraWorldForward.x * distance,
+        cameraWorldPosition.y + cameraWorldForward.y * distance,
+        cameraWorldPosition.z + cameraWorldForward.z * distance
+      );
+      transform.worldRotationQuaternion.copyFrom(cameraTransform.worldRotationQuaternion);
+    } else {
+      const { canvas } = this.engine;
+      transform.setWorldPosition(canvas.width / 2, canvas.height / 2, 0);
+      transform.worldRotationQuaternion.set(0, 0, 0, 1);
+    }
+  }
+
+  private _adapterSizeInScreenSpace(): void {
+    const { _renderCamera: renderCamera } = this;
+    const { x: width, y: height } = this._referenceResolution;
+    let curWidth: number;
+    let curHeight: number;
+    if (renderCamera) {
+      curHeight = renderCamera.isOrthographic
+        ? renderCamera.orthographicSize * 2
+        : 2 * (Math.tan(MathUtil.degreeToRadian(renderCamera.fieldOfView / 2)) * this._distance);
+      curWidth = renderCamera.aspectRatio * curHeight;
+    } else {
+      const canvas = this.engine.canvas;
+      curHeight = canvas.height;
+      curWidth = canvas.width;
+    }
+    let expectX: number, expectY: number, expectZ: number;
+    switch (this._resolutionAdaptationStrategy) {
+      case ResolutionAdaptationStrategy.WidthAdaptation:
+        expectX = expectY = expectZ = curWidth / width;
+        break;
+      case ResolutionAdaptationStrategy.HeightAdaptation:
+        expectX = expectY = expectZ = curHeight / height;
+        break;
+      case ResolutionAdaptationStrategy.BothAdaptation:
+        expectX = curWidth / width;
+        expectY = curHeight / height;
+        expectZ = (expectX + expectY) / 2;
+        break;
+      case ResolutionAdaptationStrategy.ExpandAdaptation:
+        expectX = expectY = expectZ = Math.min(curWidth / width, curHeight / height);
+        break;
+      case ResolutionAdaptationStrategy.ShrinkAdaptation:
+        expectX = expectY = expectZ = Math.max(curWidth / width, curHeight / height);
+        break;
+      default:
+        break;
+    }
+    this.entity.transform.setScale(expectX, expectY, expectZ);
+    this._transform.size.set(curWidth / expectX, curHeight / expectY);
+  }
+
+  private _walk(entity: Entity, elements: IUIElement[], depth = 0): void {
+    const components = entity._components;
+    for (let i = 0, n = components.length; i < n; i++) {
+      const component = components[i];
+      if (component.enabled && component._componentType & ComponentType.UIElement) {
+        (component as unknown as IUIElement).depth = depth++;
+        elements.push(component as unknown as IUIElement);
```

Comment:
Can this type conversion be optimized? I feel like Guolei might have comments on it.